### DISCUSSION TITLE
Tutorial - OpenMage installation using Composer

### DISCUSSION THREAD

**Discussion by addison74** [OpenMage/magento-lts#2381] [Questions & Answers]
Recently, enough ink has flowed about Composer. This discussion aims to lay the foundations of a tutorial on how to install OpenMage in a development environment and how to use it in production, how to update it using Composer. The discussion is useful because I found that not everyone uses Composer in OpenMage (I am one of them) or has used it in a project.

Intuitively, if someone asks me to install OpenMage using Composer, after installing it on the server I would go to **packagist.org** and search for the project name. Then I would run the command shows there ```composer require openmage/magento-lts```. 

If I run it in the root directory I get the following tree:

composer.json
composer.lock
root/
vendor/
-- autoload.php
-- well/
-- composer/
-- eloquent/
-- flyinmana/
-- justinrainbow/
-- magento-hackathon/
-- openmage/
---- **magento-lts/ (here is what we find in GitHub)**
-- psr/
-- symphony/

The problem I have now is that to run OpenMage in the browser I will have to write a rather long path /vendor/openmage/magento-lts/, but I can solve it by establishing the root directory up to here.

Any experience is welcome.

**Comment by fballiano** [comment]
This is a composer.json that will correctly install OM 20.0.15 in the "openmage" folder, then you'll have to point your webserver's root directory to that folder:

```
{
    "name": "fballiano/openmage",
    "type": "magento-core",
    "require": {
        "php": ">=8.0, <=8.1",
        "aydin-hassan/magento-core-composer-installer": "2.1.0",
        "magento-hackathon/magento-composer-installer": "4.0.2",
        "cweagans/composer-patches": "1.7.2",
        "openmage/magento-lts": "20.0.15"
    },
    "extra": {
        "composer-exit-on-patch-failure": true,
        "enable-patching": true,
        "magento-core-package-type": "magento-source",
        "magento-root-dir": "openmage"
    },
    "config": {
        "preferred-install": "dist",
        "optimize-autoloader": true,
        "classmap-authoritative": true,
        "allow-plugins": {
            "magento-hackathon/magento-composer-installer": true,
            "aydin-hassan/magento-core-composer-installer": true,
            "cweagans/composer-patches": true
        }
    },
    "repositories": {
        "firegento": {
            "type": "composer",
            "url": "https://packages.firegento.com"
        }
    }
}
```

took me some time to figure it out. composer's patches plugin isn't strictly necessary but I use it all the time to add custom patches (or PRs from openmage that are not merged yet).

**Comment by elidrissidev** [comment]
Actually you are not supposed to point the server to the `vendor/openmage/magento-lts` folder, but rather to the folder where the files were copied by the installer (the value of `magento-root-dir` in `composer.json`, which is set to `root` folder in your example). The files that are part of OpenMage are ignored by git and should not be modified for easier upgrade. You will then be able to add your custom extensions as usual to `<magento-root-dir>/app/code/local` or `<magento-root-dir>/app/code/community`. It is indeed confusing at first.
